File: CryBrary/Entity/Game Object/GameObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using CryEngine.Native;

namespace CryEngine.Advanced
{
    public class GameObject
    {
        #region Statics
        static GameObject()
        {
            GameObjects = new List<GameObject>();
        }

        public static GameObject Get(EntityId id)
        {
            var handle = NativeGameObjectMethods.GetGameObject(id);
            if (handle == IntPtr.Zero)
                return null;

            var gameObject = GameObjects.FirstOrDefault(x => x.Handle == handle);
            if (gameObject == null)
            {
                gameObject = new GameObject(handle);

                GameObjects.Add(gameObject);
            }

            return gameObject;
        }

        private static List<GameObject> GameObjects { get; set; }
        #endregion

        GameObject(IntPtr handle)
        {
            this.SetIGameObject(handle);

            Extensions = new List<GameObjectExtension>();
        }

		[CLSCompliant(false)]
		public bool SetAspectProfile(EntityAspects aspect, ushort profile, bool fromNetwork = false)
		{
			return NativeGameObjectMethods.SetAspectProfile(Handle, aspect, profile, fromNetwork);
		}

		public void EnablePhysicsEvent(bool enable, EntityPhysicsEvents physicsEvent)
		{
			NativeGameObjectMethods.EnablePhysicsEvent(Handle, enable, physicsEvent);
		}

		public bool WantsPhysicsEvent(EntityPhysicsEvents physicsEvent)
		{
			return NativeGameObjectMethods.WantsPhysicsEvent(Handle, physicsEvent);
		}

        /// <summary>
        /// Query extension. returns 0 if extension is not there.
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        public GameObjectExtension QueryExtension(string name)
        {
            return TryGetExtension(NativeGameObjectMethods.QueryExtension(Handle, name));
        }

        /// <summary>
        /// Forcefully get a pointer to an extension (may instantiate if needed)
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        public GameObjectExtension AcquireExtension(string name)
        {
            return TryGetExtension(NativeGameObjectMethods.AcquireExtension(Handle, name));
        }

        /// <summary>
        /// Release a previously acquired extension
        /// </summary>
        /// <param name="name"></param>
        public void ReleaseExtension(string name)
        {
            NativeGameObjectMethods.ReleaseExtension(Handle, name);
        }

        /// <summary>
        /// Turn an extension on
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        public bool ActivateExtension(string name)
        {
            return NativeGameObjectMethods.ActivateExtension(Handle, name);
        }

        /// <summary>
        /// Turn an extension off
        /// </summary>
        /// <param name="name"></param>
        public void DeactivateExtension(string name)
        {
            NativeGameObjectMethods.DeactivateExtension(Handle, name);
        }

        GameObjectExtension TryGetExtension(IntPtr handle)
        {
            if (handle == IntPtr.Zero)
                return null;

            var extension = Extensions.FirstOrDefault(x => x.Handle == handle);
            if (extension == null)
            {
                extension = new GameObjectExtension();
                extension.Handle = handle;
                extension.Owner = this;

                Extensions.Add(extension);
            }

            return extension;
        }

        public void NotifyNetworkStateChange(int aspect)
        {
            NativeGameObjectMethods.ChangedNetworkState(Handle, aspect);
        }

        public PrePhysicsUpdateMode PrePhysicsUpdateMode
        {
            set { NativeGameObjectMethods.EnablePrePhysicsUpdates(Handle, value); }
        }

        List<GameObjectExtension> Extensions { get; set; }

        internal IntPtr Handle { get; set; }
    }
}

File: MonoDll/Scriptbinds/GameObject.cpp
#include "StdAfx.h"
#include "GameObject.h"

#include <IGameObject.h>

CScriptbind_GameObject::CScriptbind_GameObject()
{
	REGISTER_METHOD(GetGameObject);

	REGISTER_METHOD(EnablePostUpdates);
	REGISTER_METHOD(EnablePrePhysicsUpdates);

	REGISTER_METHOD(QueryExtension);
	REGISTER_METHOD(AcquireExtension);

	REGISTER_METHOD(ReleaseExtension);

	REGISTER_METHOD(ActivateExtension);
	REGISTER_METHOD(DeactivateExtension);

	REGISTER_METHOD(ChangedNetworkState);

	REGISTER_METHOD(SetAspectProfile);
}

IGameObject *CScriptbind_GameObject::GetGameObject(EntityId id)
{
	return gEnv->pGameFramework->GetGameObject(id);
}

void CScriptbind_GameObject::EnablePostUpdates(IGameObject *pGameObject, IGameObjectExtension *pExtension, bool enable)
{
	if(enable)
		pGameObject->EnablePostUpdates(pExtension);
	else
		pGameObject->DisablePostUpdates(pExtension);
}

void CScriptbind_GameObject::EnablePrePhysicsUpdates(IGameObject *pGameObject, EPrePhysicsUpdate rule)
{
	pGameObject->EnablePrePhysicsUpdate(rule);
}

IGameObjectExtension *CScriptbind_GameObject::QueryExtension(IGameObject *pGameObject, mono::string name)
{
	return pGameObject->QueryExtension(ToCryString(name));
}

IGameObjectExtension *CScriptbind_GameObject::AcquireExtension(IGameObject *pGameObject, mono::string name)
{
	return pGameObject->AcquireExtension(ToCryString(name));
}

void CScriptbind_GameObject::ReleaseExtension(IGameObject *pGameObject, mono::string name)
{
	pGameObject->ReleaseExtension(ToCryString(name));
}

bool CScriptbind_GameObject::ActivateExtension(IGameObject *pGameObject, mono::string name)
{
	return pGameObject->ActivateExtension(ToCryString(name));
}

void CScriptbind_GameObject::DeactivateExtension(IGameObject *pGameObject, mono::string name)
{
	pGameObject->DeactivateExtension(ToCryString(name));
}

void CScriptbind_GameObject::ChangedNetworkState(IGameObject *pGameObject, int aspect)
{
	pGameObject->ChangedNetworkState((NetworkAspectType)aspect);
}

bool CScriptbind_GameObject::SetAspectProfile(IGameObject *pGameObject, EEntityAspects aspect, uint8 profile, bool fromNetwork)
{
	return pGameObject->SetAspectProfile(aspect, profile, fromNetwork);
}